lib/rubygems/package.rb

```diff
@@ -391,6 +393,31 @@ def gzip_to io # :yields: gz_io
   end
 
   ##
+  # Returns file mode for installing +filename+
+
+  def install_mode filename, mode
+    if
+      mode & 0111 > 0 ||
+      file_is_executable(filename)
+    then
+      0775
+    else
+      0664
+    end
+  end
+
+  ##
+  # Check if the given file is listed in specification executables list
+
+  def file_is_executable filename
+    @spec &&
+    @spec.bindir &&
+    @spec.executables &&
+    filename.start_with?(@spec.bindir) &&
+    @spec.executables.map { |exe| File.join(@spec.bindir, exe) }.include?(filename)
```

Comment:
what if instead remapping all executables we cut the file name, any performance/code quality issues with that approach?